dev/Installer/src/install.cpp

```diff
@@ -0,0 +1,172 @@
+    {
+        return true;
+    }
+
+    return false;
+}
+
+wil::com_ptr<IStream> get_resource_stream(std::wstring resourceName, std::wstring resourceType)
+{
+    HMODULE const hmodule = GetModuleHandle(NULL);
+    HRSRC hrsrc = ::FindResource(hmodule, resourceName.c_str(), resourceType.c_str());
+    THROW_LAST_ERROR_IF_NULL(hrsrc);
+    HGLOBAL hresource = LoadResource(hmodule, hrsrc);
+    THROW_LAST_ERROR_IF_NULL(hresource);
+    const BYTE* data = reinterpret_cast<BYTE*>(::LockResource(hresource));
+    THROW_LAST_ERROR_IF_NULL(data);
+    DWORD size = ::SizeofResource(hmodule, hrsrc);
+    wil::com_ptr<IStream> stream = ::SHCreateMemStream(data, size);
+    return stream;
+}
+
+std::unique_ptr<PackageProperties> get_package_properites_from_stream(wil::com_ptr<IStream>& stream)
+{
+    // Get PackageId from the manifest.
+    auto factory = wil::CoCreateInstance<AppxFactory, IAppxFactory>();
+    wil::com_ptr<IAppxPackageReader> reader;
+    THROW_IF_FAILED(factory->CreatePackageReader(stream.get(), wil::out_param(reader)));
+    wil::com_ptr<IAppxManifestReader> manifest;
+    THROW_IF_FAILED(reader->GetManifest(wil::out_param(manifest)));
+    wil::com_ptr<IAppxManifestPackageId> id;
+    THROW_IF_FAILED(manifest->GetPackageId(&id));
+
+    // Populate properties from the manifest PackageId
+    auto properties = std::make_unique<PackageProperties>();
+    THROW_IF_FAILED(id->GetPackageFullName(&properties->full_name));
+    THROW_IF_FAILED(id->GetPackageFamilyName(&properties->family_name));
+    APPX_PACKAGE_ARCHITECTURE arch;
+    THROW_IF_FAILED(id->GetArchitecture(&arch));
+    properties->architecture = static_cast<ProcessorArchitecture>(arch);
+    THROW_IF_FAILED(id->GetVersion(&properties->version));
+
+    return properties;
+}
+
+wil::com_ptr<IStream> open_temp_file(_In_ PCWSTR path)
+{
+    wil::com_ptr<IStream> outstream;
+    THROW_IF_FAILED(SHCreateStreamOnFileEx(path, STGM_WRITE | STGM_READ | STGM_SHARE_DENY_WRITE | STGM_CREATE, FILE_ATTRIBUTE_NORMAL, TRUE, nullptr, wil::out_param(outstream)));
+    return outstream;
+}
+
+winrt::hresult deploy_package_from_resource(ResourcePackageInfo resource, bool quiet) noexcept
+{
+    // Get package properties by loading the resource as a stream and reading the manifest.
+    wil::com_ptr<IStream> pkgstream;
+    std::unique_ptr<PackageProperties> properties;
+    try
+    {
+        pkgstream = get_resource_stream(resource.Id, resource.ResourceType);
+        properties = get_package_properites_from_stream(pkgstream);
+    }
+    catch (wil::ResultException ex)
+    {
+        return ex.GetErrorCode();
+    }
+
+    // Skip non-applicable architectures.
+    if (!is_architecture_applicable(properties->architecture))
+    {
+        return S_OK;
+    }
+
+
+    wchar_t package_filename[MAX_PATH];
+    RETURN_LAST_ERROR_IF(0 == GetTempFileName(std::filesystem::temp_directory_path().c_str(), L"PRP", 0u, package_filename));
```

Comment:
[External]
I would perhaps refactor at some point to utilize a uuid-based approach and one that leverages CreateFile's `FILE_FLAG_DELETE_ON_CLOSE` to make cleanup guaranteed. Would also make changes to perform the appropriate measurements to eliminate MAX_PATH.

Reply:
I tried using FILE_FLAG_DELETE_ON_CLOSE initially and ran into problems with the deployment API not using it. I suspect the issue is that for FILE_FLAG_DELETE_ON_CLOSE to work properly, all openers must open with FILE_FLAG_SHARE_DELETE, and I don't think the deployment pipeline is doing that.

I order to keep the installer simple and reliable, I'm going to avoid using this flag and simply delete the file on scope exit. It's still being created in a temporary directory so if for some reason scope exit doesn't properly delete the file it can still be reclaimed when cleaning up temporary files.